### PR TITLE
 Device Passthrough options added

### PR DESCRIPTION
**Pull request by emrahbillur**
Orin NX and Orin AGX has different device passthrough options for different cards. Selective options added for passthrough.

### CODE REVIEW

**Comment by vilvo** [comment]
I'm still in the process of testing this. Can you please fix the commit message (see below) to remove the duplicate lines and please add description of what functionality related to NIC passthrough to `netvm` can be expected to work (if any) after this change.
screencapture

**Comment by vilvo** [comment]
Optionally, if this is purely and only to pass the configured - SoM specific passthrough parameters but passthrough network not yet working, please state that in the commit message. Commit messages are relevant for any future reader of the version history.

**Comment by emrahbillur** [comment]
The double message in commit is corrected. Added relevant information on what is working on netvm and what is needed to make pci ethernet work on it in the commit message.

**Review by vilvo — APPROVED** [review]

File: targets/nvidia-jetson-orin.nix [hunk 1]
```diff
@@ -10,14 +10,25 @@
...
+  somDevices = {
+    "agx" = {
+      bus = "pci";
+      path = "0001:01:00.0";
+    };
+    "nx" = {
+      bus = "pci";
+      path = "0008:01:00.0";
+    };
+  };
```
Comment (brianmcgillion) [inline targets/nvidia-jetson-orin.nix:22]: did you look at defining these in modules/hardware/nvidia-jetson-orin/jetson-orin.nix, to keep the HW definitions self contained ?

Comment (emrahbillur) [reply targets/nvidia-jetson-orin.nix:22]: I'll try and reply the results.

Comment (emrahbillur) [reply targets/nvidia-jetson-orin.nix:22]: Unfortunately my several attempts failed to move that into modules hardware section. I suggest let's move on with this way for now while I'll try to work on that later. Ghaf on Orin Nx will be on use quite soon and there is still work to be finished to insert the firmware of ethernet.

Comment (emrahbillur) [reply targets/nvidia-jetson-orin.nix:22]: I'm just trying one last change which might work for moving all hardware options to modules. This might do the trick. If it also fails then at this step this might be the last trial for today.

Comment (emrahbillur) [reply targets/nvidia-jetson-orin.nix:22]: Finally solved moving device specific  part to modules/hardware/nvidia-jetson-orin/jetson-orin.nix on my last try. I'll commit that code now.

Comment (emrahbillur) [reply targets/nvidia-jetson-orin.nix]: Hi, I moved all hardware related options to modules/hardware/nvidia-jetson-orin/jetson-orin.nix. The changes isolated hardware options in modules part. Microvm works and this also solves the https://ssrc.atlassian.net/browse/SP-2585 issue. Formatting and all checks are done. I'll now add the ethernet firmware support for realtek 8168 so as to make netvm fully functional.

**Comment by emrahbillur** [comment]
> Would you please add the ethernet firmware missing from netvm with this PR? Due to the unresolved Orin NX debug UART TX issue on some units, the device is practically inaccessible (only debug RX out) if we merge this PR as is.

Sure. I'm still working on the firmware from the morning. To support usb wifi I'm also trying to solve and add the wifi regulatory db firmware issue with the image by adding 
```
      hardware.firmware = [ pkgs.wireless-regdb ];
      boot.extraModprobeConfig = ''
        options cfg80211 ieee80211_regdom="FI"
      '';
```
and also the realtek ethernet firmware issue is on the way.

**Comment by vilvo** [comment]
> > Would you please add the ethernet firmware missing from netvm with this PR? Due to the unresolved Orin NX debug UART TX issue on some units, the device is practically inaccessible (only debug RX out) if we merge this PR as is.
> 
> Sure. I'm still working on the firmware from the morning. To support usb wifi I'm also trying to solve and add the wifi regulatory db firmware issue with the image by adding
> 
> ```
>       hardware.firmware = [ pkgs.wireless-regdb ];
>       boot.extraModprobeConfig = ''
>         options cfg80211 ieee80211_regdom="FI"
>       '';
> ```
> 
> and also the realtek ethernet firmware issue is on the way.

`netvm` USB NIC support is secondary and must be available only with https://tiiuae.github.io/ghaf/ref_impl/modules_options.html#ghafdevelopmentdebugtoolsenable `ghaf.development.debug.tools.enable = true` and device specific USB passthroughs in the NX hardware module to support development/debugging. Beyond that, `release` target (`ghaf.development.debug.tools.enable = false`) `netvm` kernel should not have USB subsystem.

Long story short - please enable the NX on board ethernet passthrough and FW first.

**Comment by emrahbillur** [comment]
> > > Would you please add the ethernet firmware missing from netvm with this PR? Due to the unresolved Orin NX debug UART TX issue on some units, the device is practically inaccessible (only debug RX out) if we merge this PR as is.
> > 
> > 
> > Sure. I'm still working on the firmware from the morning. To support usb wifi I'm also trying to solve and add the wifi regulatory db firmware issue with the image by adding
> > ```
> >       hardware.firmware = [ pkgs.wireless-regdb ];
> >       boot.extraModprobeConfig = ''
> >         options cfg80211 ieee80211_regdom="FI"
> >       '';
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and also the realtek ethernet firmware issue is on the way.
> 
> `netvm` USB NIC support is secondary and must be available only with https://tiiuae.github.io/ghaf/ref_impl/modules_options.html#ghafdevelopmentdebugtoolsenable `ghaf.development.debug.tools.enable = true` and device specific USB passthroughs in the NX hardware module to support development/debugging. Beyond that, `release` target (`ghaf.development.debug.tools.enable = false`) `netvm` kernel should not have USB subsystem.
> 
> Long story short - please enable the NX on board ethernet passthrough and FW first.

Sure. My first aim is to make ethernet work in microvm. Still working.

**Comment by mikatammi** [comment]
Does NX Orin (or the board that has the chip) have different ethernet card than AGX Orin? As the ethernet card inside the SoC is not a PCI or USB device, but a platform device, the passthrough is more difficult. To get them working, we would first need to integrate support for the virtual BPMP, which requires modifications to both host and guest kernel, and the hypervisor

**Comment by emrahbillur** [comment]
> Does NX Orin (or the board that has the chip) have different ethernet card than AGX Orin? As the ethernet card inside the SoC is not a PCI or USB device, but a platform device, the passthrough is more difficult. To get them working, we would first need to integrate support for the virtual BPMP, which requires modifications to both host and guest kernel, and the hypervisor

No the ethernet devices are completely different. It is the Aquantia AQR113C ethernet device on Orin AGX and it is the Realtek 8168 device which is on Orin NX. In fact the Orin NX detects and was able to use the Realtek 8168 ethernet device without pci passthrough by means of the realtek 8169 kernel module. Unfortunately it gives a firmware issue for not loading the rtl_nic/r8168h_2.fw file. But it still works and gives ethernet access. When it comes to the PCI passthrough to netvm it is another issue. Now the netvm runs detects the ethernet 8168 through module of 8169 again fails to load firmware. But this time it can not detect the state of the ethernet cable. mii_tool can detect the cable up and down status, but eth-tool fails to detect the cable as with the ethernet driver. I'll figure out what will happen after the firmware is loaded.

**Comment by emrahbillur** [comment]
The Orin Nx support with netvm ethernet working. All AGX and NX related configuration moved to modules/target section and configured through flakes. Issue solved through ACS Override and adding qemu parameters. The problem was related with the Nx device tree storage and small QSPI. Will also make some simplification to the configuration after flash script is being added. 
Current status: Netvm Ethernet up and working, host has no ethernet, AGX and NX configuration completely moved to modules section. No flash script available yet. Creates the nixos.img and boots from USB.

**Comment by vilvo** [comment]
> ... Creates the nixos.img and boots from USB.

Fails to build with:
```
❯ nix build .#packages.aarch64-linux.nvidia-jetson-orin-nx-debug
error: build of '/nix/store/hg4bai6azyf2jnqif1h17hp8bjbaxflf-nixos-disk-image.drv' on 'ssh://vilvo@192.168.1.149' failed: builder for '/nix/store/hg4bai6azyf2jnqif1h17hp8bjbaxflf-nixos-disk-image.drv' failed with exit code 1;
       last 10 log lines:
       > tune2fs 1.46.6 (1-Feb-2023)
       > Setting maximal mount count to -1
       > Setting interval between checks to 0 seconds
       > Setting time filesystem last checked to Mon Aug 21 08:38:23 2023
       >
       > tune2fs 1.46.6 (1-Feb-2023)
       > Setting time filesystem last checked to Thu Jan  1 00:00:00 1970
       >
       > [   16.085119] reboot: Power down
       > mv: error writing '/nix/store/n8mhccwa9r2lva23w2fgifal8k8cwhzh-nixos-disk-image/nixos.img': No space left on device
       For full logs, run 'nix log /nix/store/hg4bai6azyf2jnqif1h17hp8bjbaxflf-nixos-disk-image.drv'.
       note: build failure may have been caused by lack of free disk space
error: builder for '/nix/store/hg4bai6azyf2jnqif1h17hp8bjbaxflf-nixos-disk-image.drv' failed with exit code 1;
       last 10 log lines:
       > tune2fs 1.46.6 (1-Feb-2023)
       > Setting maximal mount count to -1
       > Setting interval between checks to 0 seconds
       > Setting time filesystem last checked to Mon Aug 21 08:38:23 2023
       >
       > tune2fs 1.46.6 (1-Feb-2023)
       > Setting time filesystem last checked to Thu Jan  1 00:00:00 1970
       >
       > [   16.085119] reboot: Power down
       > mv: error writing '/nix/store/n8mhccwa9r2lva23w2fgifal8k8cwhzh-nixos-disk-image/nixos.img': No space left on device
       For full logs, run 'nix log /nix/store/hg4bai6azyf2jnqif1h17hp8bjbaxflf-nixos-disk-image.drv'.
```
Not out of disk space on my build host or remote `aarch64` builder. It's probably the disk image creation process.

**Comment by emrahbillur** [comment]
I'll try the build myself and try to find someone else to try this build also. I'll check the logs of both. 
(Note: I've also added some new part that I've not merged in this pull request which disables weston and demo apps as it was one of the jira tasks on me as a next step, where if the device is nx it does not include weston and demo apps) I'll merge it in another pull request.

**Comment by emrahbillur** [comment]
Hi Ville, I tested nix build on my machine and Berk also had the same build without any issues. There should be some temporary problem on your side.

**Comment by vilvo** [comment]
> Hi Ville, I tested nix build on my machine and Berk also had the same build without any issues. There should be some temporary problem on your side.

The build had finished on `awsarm` so yes.

**Comment by vilvo** [comment]
Functionally the NX ethernet passthrough to netvm works:
- `netvm` loads the native driver `r8169` (passthrough enabled)
- `netvm` `enp0s4f0`-interface gets the IP address via DHCP
- debug-image allows login to `netvm` over ssh with authentication module details
- debug-image allows login to `ghaf-host` from `netvm` via `virbr0`
- debug-image allows login from `ghaf-host` to `netvm` via `virbr0`

File: modules/hardware/nvidia-jetson-orin/jetson-orin.nix [hunk 2]
```diff
@@ -14,14 +16,61 @@
...
+      buspath = [
+        {
+          bus = "pci";
+          path = "0001:01:00.0";
+        }
+      ];
+      kernelParams = [];
+      wireless = {
+        enable = true;
+        networks."SSID_OF_NETWORK".psk = "WPA_PASSWORD";
```
Comment (vilvo) [inline modules/hardware/nvidia-jetson-orin/jetson-orin.nix:28]: This should not be left enabled (uncommented) as it will try to configure SSID: `SSID_OF_NETWORK` with password: `WPA_PASSWORD`. Please leave commented out or refactor into a WIFI module that can be enabled/disabled.

File: modules/hardware/nvidia-jetson-orin/jetson-orin.nix [hunk 3]
```diff
@@ -14,14 +16,61 @@
...
+        enable = false;
+      };
     };
   };
+  netvmExtraModules = [
+    {
+      # This part should be conditional for AGX 01:01 for NX 08:01
+      microvm.devices = somDefinition."${cfg.somType}".buspath;
+
+      # nixpkgs.config.allowUnfree = true;
```
Comment (vilvo) [inline modules/hardware/nvidia-jetson-orin/jetson-orin.nix:63]: Is there a reason to `allowUnfree = false`? If yes, please remove commented out.

**Review by vilvo — CHANGES_REQUESTED** [review]
Please address the comments and add credits to the original work of the adapted patch.

**Comment by vilvo** [comment]
Any reason you closed this?
